Sources/SwiftFusion/MCMC/TransitionKernel.swift

```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Minimal requirements to efficiently implement a Markov chain Monte Carlo (MCMC) transition kernel. A transition kernel returns a new state given some old state. It also takes (and returns) "side information" which may be used for debugging or optimization purposes (i.e, to "recycle" previously computed results).
+/// Inspired by tpf [TransitionKernel](https://www.tensorflow.org/probability/api_docs/python/tfp/mcmc/TransitionKernel)
+public protocol TransitionKernel {
+  associatedtype State
+  associatedtype Results
+  
+  /// Takes one step of the TransitionKernel.
+  func one_step(_ current_state: State, _ previous_kernel_results: Results) -> (State, Results)
+  
+  /// Returns an object with the same type as returned by one_step(...)[1].
+  func bootstrap_results(_ init_state: State) -> Results
+}
```

Comment:
As a direct translation of the Python API, this looks right to me. If you want to go this direction, then that sounds good to me!

---

The "Swifty" way can make some things a lot nicer so I'll describe a "Swifty" idea here to tempt you :)

The `TransitionKernel` looks like an infinite [Sequence](https://developer.apple.com/documentation/swift/sequence). So there's no need to define a `protocol TransitionKernel` -- you can conform things to `Sequence` instead.

For example, `RandomWalkMetropolis` would be something like:

```swift
struct RandomWalkMetropolis<State>: Sequence, IteratorProtocol {

  var previousLogProb: Double
  var currentState: State

  let target_log_prob_fn: (State) -> Double
  let new_state_fn : (State)->State

  init(_ target_log_prob_fn: @escaping (State) -> Double, _ new_state_fn : @escaping (State)->State, _ initialState: State) {
    ...
  }

  mutating func next() -> State {
    let result = currentState

    // update `currentState` and `previousLogProb` according to metropolis rules
    ...

    return result
  }
}
```

This is nice because it allows us to use all the `Sequence` algorithms. For example, `sampleChain` becomes something like:

```swift
Array(sequence.dropFirst(num_burnin_steps).prefix(num_results))
```

Reply:
That is nice, lazy haskell style :-)

I created an issue so we can revisit, chat about larger issue of tfp interop in general.